Lockdir encode/decode roundtrip tests

gridbugs opened the pull request:
This adds some tests that a `Lock_dir.t` can be written do disk and then read back from disk to obtain a `Lock_dir.t` that matches the original.

gridbugs committed:
Add missing cases to Action.map_string_with_vars

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Implement equal and to_dyn for some types

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Add function for reading Lock_dir.t from lockdir

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Roundtrip tests for Lock_dir.t

This tests that we can take a Lock_dir.t, encode it to a lockdir on
disk, and then decode the lockdir into a Lock_dir.t that matches the
original.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Fix encoding of Lock_dir.Source.t

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

Alizter commented on src/dune_lang/action.ml:
```diff
@@ -60,7 +60,35 @@ module File_perm = struct
+
+    let equal a b =
+      match (a, b) with
+      | Eq, Eq -> true
+      | PlusEq, PlusEq -> true
+      | EqPlus, EqPlus -> true
+      | ColonEq, ColonEq -> true
+      | EqColon, EqColon -> true
+      | EqPlusEq, EqPlusEq -> true
```
You can merge these all into a single ->

Alizter commented on src/dune_pkg/lock_dir.ml:
```diff
@@ -9,6 +9,27 @@ module Source = struct
         ; checksum : (Loc.t * Checksum.t) option
         }
 
+  let equal a b =
```
Maybe this should be called equal_no_locs or something since we are explicitly ignoring the loc. Also what is the loc from/used for?

rgrinberg replied:
You could also just add a `drop_loc` function to convert all the locations into `Loc.none`. This is the convention we use elsewhere.

gridbugs replied:
@Alizter the loc is used in error messages. Will update to have a `drop_loc` function.

Alizter commented on src/dune_lang/action.ml:
```diff
@@ -60,38 +60,71 @@ module File_perm = struct
-    | EqColon
-    | EqPlusEq
+  module Op = struct
+    type t =
+      | Eq
+      | PlusEq
+      | EqPlus
+      | ColonEq
+      | EqColon
+      | EqPlusEq
```
Could you also add comments on their meaning?

rgrinberg committed:
refactor: remove all suspicious equalities

Make all equalities respect locations and add [remove_locs] functions to
have comparison that ignores locations.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg commented:
Thanks. I made all `equal` function respect locations and added `remove_locs` to the callers. Otherwise, it's a little too easy to use the wrong equality.
